### ISSUE TITLE
Make it easier to write alternative parameter renderers

### ISSUE THREAD

**Issue by akdor1154** [akdor1154/python-csql#10]
I needed to custom render parameters to adapt my sql to go into a udtf, i found this harder than it should have been.

In a perfect world I think this is a time for typeclasses but i feel like in Python that would be like trying to row a canoe up a mountain path. Maybe remove all the classmethod level stuff from the renders then can just rely on normal generics? Try out protocols? Just sigh and type any?

What I had to write:
```py
import csql.renderer.parameters as crp
from collections.abc import Collection as CollectionABC
from typing import *

class RendererState(NamedTuple):
	renderedKeys: Dict[int, crp.RenderedParameter]

class RenderedUDFParameter(NamedTuple):
	param: crp.RenderedParameter
	nextState: RendererState

class UDFParameterRenderer:
	# renders parameters as just their key. Useful for turning a query into a UDF. e.g. `select * from t where name = name`.

	@classmethod
	def _renderScalar(Self, paramKey, paramValue):
		return crp.RenderedParameter(
			sql=paramKey,
			values=[]
		)
	@classmethod
	def render(Self, paramKey: str, parameters: Parameters, state: RendererState) -> RenderedUDFParameter:
		paramValue = parameters.params[paramKey]

		key = id(parameters) ^ hash(paramKey)

		if key in state.renderedKeys:
			preRendered = state.renderedKeys[key]
			return RenderedUDFParameter(
				param=preRendered,
				nextState=state
			)

		if isinstance(paramValue, CollectionABC) and not isinstance(paramValue, str):
			raise NotImplementedError()
		else:
			param = Self._renderScalar(paramKey, paramValue)

		return RenderedUDFParameter(
			param=param,
			nextState=RendererState(
				renderedKeys = {**state.renderedKeys, key: param}
			)
		)

	@classmethod
	def initialState(Self) -> RendererState:
		return RendererState({})

class UDFSQLRenderer(csql.renderer.query.BoringSQLRenderer):
	paramRenderer = UDFParameterRenderer

UDFSQLRenderer.render(qSessionsAggregate)
```

**Comment by akdor1154** [comment]
nearly done in #18, probably just need to let people pass their own class in instead of enum membership in SQLDialect

**Comment by akdor1154** [comment]
fixed completely in #21
